### DISCUSSION TITLE
How to properly test something that relies on NetworkManager service?

### DISCUSSION THREAD

**Discussion by souliane** [python-sdbus/python-sdbus-networkmanager#61] [Q&A]
Hi,
I would like to use this project to serve the main features of NetworkManager via an API, and I am wondering about the best way to test it. I typically need to test CRUD operations on the connection profiles. Testing directly on the real system bus is not an option. If you were me, what would you do?

1. use `IsolatedDbusTestCase` provided by sdbus: if I understand correctly, this spawn a new _session_ bus server. Is there an easy way to replicate dbus NetworkManager to a session bus? Would that setup a trustable testing environment?
2. use https://github.com/martinpitt/python-dbusmock: I find it very convenient (it allows to mock the system bus), but it's not covering 100% of my needs, the mock might not behave exactly like the real dbus service, and extending it is cumbersome.

Thanks for any input!

**Comment by igo95862** [comment]
Hello @souliane

What you can potentially do is use the NetworkManager interface classes defined in this library to create your own objects. In python-sdbus the client and server interfaces are the same class. So using the `@dbus_method_async_override` you can change the `raise NotImplementedError` behavior to anything you want.

This will be functionally equivalent to the `dbusmock` so it carries same issues that it does not test the real NetworkManager.

What I would recommend is running a Python type checker like `mypy --strict`. python-sdbus has a full support of type hints. The connection profiles dataclasses were generated from the NetworkManager metadata and also are fully type hinted.

I think for testing against real NetworkManager you would probably need a virtual machine. (or a real one dedicated for testing)

**Reply by souliane** [reply]
I will give it a try, thanks!

**Reply by souliane** [reply]
I don't know what I am doing wrong... if you have an idea :-)

```python
from __future__ import annotations

from sdbus import request_default_bus_name_async
from sdbus.unittest import IsolatedDbusTestCase
from sdbus_async.networkmanager.objects import NETWORK_MANAGER_SERVICE_NAME
from sdbus_async.networkmanager.objects import NetworkManager


class TestInit(IsolatedDbusTestCase):
    async def test(self) -> None:
        await request_default_bus_name_async(NETWORK_MANAGER_SERVICE_NAME, 0)

        path = "/org/freedesktop/NetworkManager"
        NetworkManager().export_to_dbus(path)

        proxy = NetworkManager.new_proxy(NETWORK_MANAGER_SERVICE_NAME, path)
        await proxy.version

```
It fails with:
```python
>       await proxy.version

tests/test_init.py:17: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <sdbus.dbus_proxy_async_property.DbusPropertyAsyncBinded object at 0x7f9ec857a390>

    async def get_async(self) -> T:
        assert self.interface_ref is not None
        interface = self.interface_ref()
        assert interface is not None
    
        if not interface._is_binded:
            return self.dbus_property.property_getter(
                interface)
    
        assert interface._attached_bus is not None
        assert interface._remote_service_name is not None
        assert interface._remote_object_path is not None
        assert self.dbus_property.property_name is not None
        assert self.dbus_property.interface_name is not None
        new_call_message = interface._attached_bus. \
            new_property_get_message(
                interface._remote_service_name,
                interface._remote_object_path,
                self.dbus_property.interface_name,
                self.dbus_property.property_name,
            )
    
>       reply_message = await interface._attached_bus. \
            call_async(new_call_message)
E       sdbus.dbus_exceptions.DbusUnknownObjectError: Unknown object '/org/freedesktop/NetworkManager'.

../../.cache/pypoetry/virtualenvs/network-manager-taguGb95-py3.11/lib/python3.11/site-packages/sdbus/dbus_proxy_async_property.py:150: DbusUnknownObjectError
```

**Reply by igo95862** [reply]
There are two problems in your code.

First, you are using the `NetworkManager` class from `sdbus_async.networkmanager.objects`. The issue is the classes from that file are intended to be used with real NetworkManager and automatically proxy themselves.

What you want to use are the raw interface classes such as `NetworkManagerInterfaceAsync` from `sdbus_async.networkmanager.interfaces_other`. This class can be subclassed, exported and proxied.

Second problem is that you instantly destroy the exported object because you have not assigned it to a variable. So when you call the proxy the object won't exist anymore.

I fixed those two issues and this code raises the NotImplementedError as expected:

```python
from __future__ import annotations

from unittest import main

from sdbus import request_default_bus_name_async
from sdbus.unittest import IsolatedDbusTestCase
from sdbus_async.networkmanager.objects import NETWORK_MANAGER_SERVICE_NAME
from sdbus_async.networkmanager.interfaces_other import NetworkManagerInterfaceAsync


class TestInit(IsolatedDbusTestCase):
    async def test(self) -> None:
        await request_default_bus_name_async(NETWORK_MANAGER_SERVICE_NAME, 0)

        path = "/org/freedesktop/NetworkManager"
        nm_mock = NetworkManagerInterfaceAsync()
        nm_mock.export_to_dbus(path)

        proxy = NetworkManagerInterfaceAsync.new_proxy(
            NETWORK_MANAGER_SERVICE_NAME,
            path,
        )
        await proxy.version


if __name__ == "__main__":
    main()
```

**Reply by igo95862** [reply]
Using the `patch` from `unittest.mock` you can change the behavior of interfaces without overriding methods:

```python
        with patch.object(
            NetworkManagerInterfaceAsync.version.dbus_property,
            "property_getter",
            return_value="unittest",
        ) as version_mock:
            nm_mock.export_to_dbus(path)
            self.assertEqual(
                await proxy.version,
                "unittest",
            )

        version_mock.assert_called()
```
Although this kind of relies on undocumented API parts of python-sdbus which will get changed in next version.

**Comment by souliane** [comment]
I realize that I should not need this for testing network manager, it looks more suitable for testing my own services.
If I do this, I would add some unneeded complexity (re-test sdbus...), and I better mock sdbus client methods, not the server ones. So I also don't need to use `IsolatedDbusTestCase` because my tests will run without dbus daemon.
It's good to know what I can do with `IsolatedDbusTestCase` though - thanks for explaining me :-)

> Second problem is that you instantly destroy the exported object because you have not assigned it to a variable. So when you call the proxy the object won't exist anymore.

Oh yeah, I remember reading this in the documentation :-p

**Reply by igo95862** [reply]
I would still recommend running the type checker. It will catch a lot of errors basically without any investments.

**Reply by souliane** [reply]
That for sure :-)
